crypto/x509/x509_vfy.c

```diff
@@ -3143,14 +3146,18 @@ static int build_chain(X509_STORE_CTX *ctx)
                  * trusted matching issuer.  Otherwise, grow the chain.
                  */
                 if (!self_signed) {
-                    curr = issuer;
-                    if ((self_signed = X509_self_signed(curr, 0)) < 0)
-                        goto int_err;
-                    if (!sk_X509_push(ctx->chain, curr)) {
+                    if (!sk_X509_push(ctx->chain, issuer)) {
                         X509_free(issuer);
                         goto memerr;
                     }
-                } else if (num == ctx->num_untrusted) {
+                    if ((self_signed = X509_self_signed(issuer, 0)) < 0)
+                        goto int_err;
+                } else {
+                    /* This should be the first self-signed trusted cert */
+                    if (!ossl_assert(ctx->num_untrusted == num)) {
```

Comment:
This assertion looks wrong.  Yes, there should only be one self-signed trusted cert, but there could be other trusted certificates that are not self-signed, and yet the chain construction may have needed to continue further.

Reply:
This branch of the code should only have been hit subject to the original condition `num == ctx->num_untrusted` which is being removed in error.

Reply:
@DDvO ping?

Reply:
> This assertion looks wrong. Yes, there should only be one self-signed trusted cert, but there could be other trusted certificates that are not self-signed, and yet the chain construction may have needed to continue further.

Oh right - so I've removed this assertion.

Reply:
@vdukhovni, do you now agree on this?
Thanks for your invitation for having a phone call on the various open points regarding cert (chain) validation that you wrote elsewhere.
Yet unfortunately I won't have time for diving into this topic any further before my upcoming vacation.

Reply:
Ping @vdukhovni - can you give your blessing on this PR now?